envbuilder.go

```diff
@@ -586,117 +652,130 @@ func Run(ctx context.Context, opts options.Options) error {
 			for _, container := range devContainer {
-				if container.RemoteUser != "" {
-					opts.Logger(log.LevelInfo, "#%d: 🧑 Updating the user to %q!", stageNumber, container.RemoteUser)
+				if container.ContainerUser != "" {
+					opts.Logger(log.LevelInfo, "#%d: 🧑 Updating the user to %q!", stageNumber, container.ContainerUser)
 
-					configFile.Config.User = container.RemoteUser
+					configFile.Config.User = container.ContainerUser
 				}
-				maps.Copy(containerEnv, container.ContainerEnv)
-				maps.Copy(remoteEnv, container.RemoteEnv)
+				maps.Copy(runtimeData.ContainerEnv, container.ContainerEnv)
+				maps.Copy(runtimeData.RemoteEnv, container.RemoteEnv)
 				if !container.OnCreateCommand.IsEmpty() {
-					scripts.OnCreateCommand = container.OnCreateCommand
+					runtimeData.Scripts.OnCreateCommand = container.OnCreateCommand
 				}
 				if !container.UpdateContentCommand.IsEmpty() {
-					scripts.UpdateContentCommand = container.UpdateContentCommand
+					runtimeData.Scripts.UpdateContentCommand = container.UpdateContentCommand
 				}
 				if !container.PostCreateCommand.IsEmpty() {
-					scripts.PostCreateCommand = container.PostCreateCommand
+					runtimeData.Scripts.PostCreateCommand = container.PostCreateCommand
 				}
 				if !container.PostStartCommand.IsEmpty() {
-					scripts.PostStartCommand = container.PostStartCommand
+					runtimeData.Scripts.PostStartCommand = container.PostStartCommand
 				}
 			}
 		}
 
-		// Sanitize the environment of any opts!
-		options.UnsetEnv()
-
-		// Remove the Docker config secret file!
-		if err := cleanupDockerConfigJSON(); err != nil {
-			return err
+		maps.Copy(runtimeData.ContainerEnv, buildParams.ContainerEnv)
+		maps.Copy(runtimeData.RemoteEnv, buildParams.RemoteEnv)
+		if runtimeData.ContainerUser == "" && configFile.Config.User != "" {
+			runtimeData.ContainerUser = configFile.Config.User
 		}
-
-		environ, err := os.ReadFile("/etc/environment")
+	} else {
+		devcontainerPath, _, err := findDevcontainerJSON(opts.WorkspaceFolder, opts)
 		if err == nil {
-			for _, env := range strings.Split(string(environ), "\n") {
-				pair := strings.SplitN(env, "=", 2)
-				if len(pair) != 2 {
-					continue
+			file, err := opts.Filesystem.Open(devcontainerPath)
+			if err != nil {
+				return fmt.Errorf("open devcontainer.json: %w", err)
+			}
+			defer file.Close()
+			content, err := io.ReadAll(file)
+			if err != nil {
+				return fmt.Errorf("read devcontainer.json: %w", err)
+			}
+			devContainer, err := devcontainer.Parse(content)
+			if err == nil {
+				maps.Copy(runtimeData.ContainerEnv, devContainer.ContainerEnv)
+				maps.Copy(runtimeData.RemoteEnv, devContainer.RemoteEnv)
+				if devContainer.ContainerUser != "" {
+					runtimeData.ContainerUser = devContainer.ContainerUser
 				}
-				os.Setenv(pair[0], pair[1])
+				runtimeData.Scripts = devContainer.LifecycleScripts
+			} else {
+				opts.Logger(log.LevelError, "Failed to parse devcontainer.json: %s", err.Error())
 			}
```

Comment:
Review: We're parsing the _workspace_ `devcontainer.json` here as we don't have access to the build-time `devcontainer.json`.

Motivation: Unless we cache the whole repo in the image, we can't reliably use `devcontainer.json` from it, it could be referencing unknown files from the repo which we haven't cached.